internal/testclient/testtx/context.go

```diff
@@ -0,0 +1,350 @@
+		t, signingKeyName,
+		keyring,
+		expectedTx,
+		expectedTxHash,
+	)
+
+	// intercept #BroadcastTxSync() call to mock response and prevent actual broadcast
+	txCtxMock.EXPECT().BroadcastTxSync(gomock.Any()).
+		DoAndReturn(func(txBytes []byte) (*cosmostypes.TxResponse, error) {
+			return &cosmostypes.TxResponse{
+				Height:    1,
+				TxHash:    expectedTxHash.String(),
+				RawLog:    "",
+				Logs:      nil,
+				Tx:        nil,
+				Timestamp: "",
+				Events:    nil,
+			}, nil
+		}).Times(1)
+
+	return txCtxMock
+}
+
+// NewBaseTxContext establishes a foundational mock transaction context with
+// predefined behaviors suitable for a broad range of testing scenarios. It ensures
+// that when interactions like transaction building, signing, and encoding occur
+// in the test environment, they produce predictable and controlled outcomes.
+//
+// Parameters:
+// - t: The testing.T instance used for the current test, typically passed from
+// the calling test function.
+// - signingKeyName: The name of the key within the keyring to be used for
+// transaction signing.
+// - keyring: The Cosmos SDK keyring containing the available cryptographic keys.
+// - expectedTx: A pointer whose value will be set to the expected transaction
+// bytes (in hexadecimal format).
+// - expectedTxHash: A pointer whose value will be set to the expected
+// transaction hash.
+// - expectedErrMsg: A pointer whose value will be set to the expected error
+// message string.
+//
+// The function works as follows:
+//  1. Invokes the NewAnyTimesTxTxContext to create a base mock transaction context.
+//  2. Sets the expectation that NewTxBuilder method will be called exactly once.
+//  3. Configures the mock behavior for the SignTx method to utilize the context's
+//     signing logic.
+//  4. Overrides the EncodeTx method's behavior to intercept the encoding operation,
+//     capture the encoded transaction bytes, compute the transaction hash, and populate
+//     the expectedTx and expectedTxHash parameters accordingly.
+func NewBaseTxContext(
+	t *testing.T,
+	signingKeyName string,
+	keyring cosmoskeyring.Keyring,
+	expectedTx *cometbytes.HexBytes,
+	expectedTxHash *cometbytes.HexBytes,
+) *mockclient.MockTxContext {
+	t.Helper()
+
+	txCtxMock, txCtx := NewAnyTimesTxTxContext(t, keyring)
+	txCtxMock.EXPECT().NewTxBuilder().
+		DoAndReturn(txCtx.NewTxBuilder).
+		Times(1)
+	txCtxMock.EXPECT().SignTx(
+		gomock.Eq(signingKeyName),
+		gomock.AssignableToTypeOf(txCtx.NewTxBuilder()),
+		gomock.Eq(false), gomock.Eq(false),
+	).DoAndReturn(txCtx.SignTx).Times(1)
+	txCtxMock.EXPECT().EncodeTx(gomock.Any()).
+		DoAndReturn(func(txBuilder cosmosclient.TxBuilder) (_ []byte, err error) {
+			// intercept cosmosTxContext#EncodeTx to get the encoded tx cometbytes
+			*expectedTx, err = txCtx.EncodeTx(txBuilder)
+			*expectedTxHash = comettypes.Tx(expectedTx.Bytes()).Hash()
+			require.NoError(t, err)
+			return expectedTx.Bytes(), nil
+		}).Times(1)
```

Comment:
Given that the function name does not include `OneTime`, is it fair to assume `.Times(1`) everywhere?

Reply:
Nice catch! :raised_hands:  I think this can be `AnyTimes()` for now.